### PR TITLE
virts-2892: Knowledge Service Documentation

### PR DESCRIPTION
**Pull request by ArtificialErmine**
## Description

Adds some documentation regarding the Knowledge Service to "How to Build Planners".

## Type of change
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## How Has This Been Tested?

It's just documentation changes, so I technically haven't tested this. The core code  being documented has been in use for a while, though, so it's pretty tested at this point.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [NA] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: sphinx-docs/How-to-Build-Planners.md [hunk 1]
```diff
@@ -196,3 +196,41 @@ Within a planner, all public utilities are available from `self.operation`. The
...
+```python
+app.objects.secondclass.c_rule
+```
+| method | description |
+|:-------|:------------|
+|`KnowledgeService.add_rule(rule)` | Add a rule to the datastore. The `rule` argument must be an already existing `Rule()` object.|
+|`KnowledgeService.get_rules(criteria)` | Retrieve matching rules from the datastore. The `criteria` argument should be a dictionary containing files to match existing rules against.|
+|`KnowledgeService.delete_rule(criteria)` | Remove a rule from the datastore. The `criteria` argument should be a dictionary containing fields and values to match existing rules against.|
+
+All objects added to the Knowledge service are checked against existing objects in order to enforce de-duplication, with one caveat. As origin is tracked for facts generated by links in the current implementation, this means duplicate facts created during different operations can exist in the fact store simultaneously. Facts/Relationships are usually automatically added to the fact service by `Link` objects as part of the process of parsing output, though they can be added manually should the need arise.
```
Comment (clenk) [inline sphinx-docs/How-to-Build-Planners.md:236]:
```suggestion
All objects added to the Knowledge service are checked against existing objects in order to enforce de-duplication, with one caveat. As origin is tracked for facts generated by links in the current implementation, this means duplicate facts created during different operations can exist in the fact store simultaneously. Facts/Relationships are usually automatically added to the fact store by `Link` objects as part of the process of parsing output, though they can be added manually should the need arise.
```

**Review by clenk — CHANGES_REQUESTED** [review]
Maybe additional details should be added to https://caldera.readthedocs.io/en/latest/Basic-Usage.html#facts as well?

Stuff that would be helpful to document somewhere:
 - fact properties other than name/value/score (already explained on the Basic Usage page)
 - the different origin types and what they mean

We should update any instances of a fact's `trait` in the docs to `name`.

I would make the order of methods consistent between fact, relationship, and rule. Currently the fact table starts with the `add_` method but the relationship one starts with the `get_` method.

Also, and I know it's not tied to this ticket, but there's a typo in the second sentence under the `Operation Utilities` section: `follow` should be `following`.

File: sphinx-docs/How-to-Build-Planners.md [hunk 2]
```diff
@@ -196,3 +196,41 @@ Within a planner, all public utilities are available from `self.operation`. The
...
 * `all_relationships()` - Return a list of all relationships collected during an operation.
 * `active_agents()` - Find all agents in the operation that have been active since operation start.
+
+## Knowledge Service
+
+As of Caldera V4.0, a new service has been added to the core of Caldera for use with planners and other components that make use of facts: the Knowledge Service. This service provides a variant of the functionality present in the object store utilized throughout Caldera, specifically tailored to facts, relationships, and rules.
+Typically, users should not need to interact with this service directly, as common usage patterns are already baked into core objects such as `Link()`, `Agent()`, and `Operation()`, but the service can be accessed by using `BaseService.get_service('knowledge_svc')`, should the need arise for more complex interactions with the available data.
+The Knowledge Service stores data persistently in the same manner that Caldera's internal Data sService does (by writing it to a file on shutdown), and can be cleared in much the same way if necessary (by using the `--fresh` argument on the server).
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply sphinx-docs/How-to-Build-Planners.md:204]: Thanks for catching that.

File: sphinx-docs/How-to-Build-Planners.md [hunk 3]
```diff
@@ -196,3 +196,41 @@ Within a planner, all public utilities are available from `self.operation`. The
...
+
+The following methods are present and supported in the Knowledge Service:
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply sphinx-docs/How-to-Build-Planners.md:206]: Makes sense.

File: sphinx-docs/How-to-Build-Planners.md [hunk 4]
```diff
@@ -196,3 +196,41 @@ Within a planner, all public utilities are available from `self.operation`. The
 * `has_fact()` - Search an operation for a fact with a particular trait and value.
 * `all_relationships()` - Return a list of all relationships collected during an operation.
 * `active_agents()` - Find all agents in the operation that have been active since operation start.
+
+## Knowledge Service
+
+As of Caldera V4.0, a new service has been added to the core of Caldera for use with planners and other components that make use of facts: the Knowledge Service. This service provides a variant of the functionality present in the object store utilized throughout Caldera, specifically tailored to facts, relationships, and rules.
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply sphinx-docs/How-to-Build-Planners.md:202]: That would be more straightforward, so I'll swap things over. The former would have been potentially useful for developers, which is why I original went with that, but it should be fairly obvious anyway.

File: sphinx-docs/How-to-Build-Planners.md [hunk 5]
```diff
@@ -196,3 +196,41 @@ Within a planner, all public utilities are available from `self.operation`. The
...
+Typically, users should not need to interact with this service directly, as common usage patterns are already baked into core objects such as `Link()`, `Agent()`, and `Operation()`, but the service can be accessed by using `BaseService.get_service('knowledge_svc')`, should the need arise for more complex interactions with the available data.
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply sphinx-docs/How-to-Build-Planners.md:203]: It was mainly to denote the referenced things were literal objects. If you don't think they're necessary I'll drop them.

File: sphinx-docs/How-to-Build-Planners.md (hunk 1 above)
Comment (ArtificialErmine) [reply sphinx-docs/How-to-Build-Planners.md:236]: That's definitely more clear.

File: sphinx-docs/How-to-Build-Planners.md [hunk 6]
```diff
@@ -196,3 +196,41 @@ Within a planner, all public utilities are available from `self.operation`. The
...
+
+```python
+app.objects.secondclass.c_rule
+```
+| method | description |
+|:-------|:------------|
+|`KnowledgeService.add_rule(rule)` | Add a rule to the datastore. The `rule` argument must be an already existing `Rule()` object.|
+|`KnowledgeService.get_rules(criteria)` | Retrieve matching rules from the datastore. The `criteria` argument should be a dictionary containing files to match existing rules against.|
+|`KnowledgeService.delete_rule(criteria)` | Remove a rule from the datastore. The `criteria` argument should be a dictionary containing fields and values to match existing rules against.|
+
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply sphinx-docs/How-to-Build-Planners.md:235]: We never implemented one - rules shouldn't be changed nearly as often as the other objects in the knowledge service. If we see a use case where changing them would be helpful in the future, I think it wouldn't be too hard to add in, but until then, I'm inclined to leave things as they are.

**Comment by ArtificialErmine** [comment]
The tables might not be working in the generated version - can you verify that they do on your end, @clenk? They work fine in markdown, but the generated html is having trouble with it on my end (and based on your previous comment it's possible that they worked fine in the html for you),

**Comment by ArtificialErmine** [comment]
> The tables might not be working in the generated version - can you verify that they do on your end, @clenk? They work fine in markdown, but the generated html is having trouble with it on my end (and based on your previous comment it's possible that they worked fine in the html for you),

I've removed the tables and reverted to using a bulleted list. The built-in triggers for generating tables in Sphinx don't appear to be working, and rendering standard Markdown tables doesn't seem to be possible without more libraries, which I'd rather avoid using. It does look way uglier this way, though.

**Review by clenk — APPROVED** [review]
Looks good, checked it locally on my end. Using bulleted lists keeps it consistent with the rest of the page.

There is a way to do tables without any additional Sphinx plugins. The markdown parser we're already using, Recommonmark, comes with something called [AutoStructify](https://recommonmark.readthedocs.io/en/latest/index.html#autostructify) that converts markdown to rst before converting to HTML. It has an option, `enable_eval_rst` that lets you stick rst in your markdown. But rst tables won't show up right if viewing the markdown in an IDE or on GitHub, and are clunkier to maintain (have to do the line wrapping manually).

At the same time, looks like Recommonmark is no longer maintained and [ReadTheDocs encourages](https://github.com/readthedocs/recommonmark/issues/221) moving to the Myst parser instead. And the Myst parser supports markdown tables.

But that's a bigger issue than this PR, which is good to go in my book. :)

**Pull request merged**